lib/Service/ImportService.php

```diff
@@ -0,0 +1,176 @@
+<?php
+
+
+namespace OCA\Tables\Service;
+
+use OC\User\NoUserException;
+use OCA\Tables\Db\Column;
+use OCA\Tables\Errors\InternalError;
+use OCA\Tables\Errors\NotFoundError;
+use OCA\Tables\Errors\PermissionError;
+use OCA\Tables\Service\ColumnTypes\TextLineBusiness;
+use OCP\DB\Exception;
+use OCP\Files\IRootFolder;
+use OCP\Files\NotFoundException;
+use OCP\Files\NotPermittedException;
+use OCP\IConfig;
+use OCP\Server;
+use PhpOffice\PhpSpreadsheet\IOFactory;
+use PhpOffice\PhpSpreadsheet\Worksheet\Row;
+use PhpOffice\PhpSpreadsheet\Worksheet\Worksheet;
+use Psr\Container\ContainerExceptionInterface;
+use Psr\Container\NotFoundExceptionInterface;
+use Psr\Log\LoggerInterface;
+
+class ImportService extends SuperService {
+
+	private IRootFolder $rootFolder;
+	private ColumnService $columnService;
+	private RowService $rowService;
+
+	private int $tableId = -1;
+	private array $columns = [];
+	private bool $createUnknownColumns = true;
+	private int $countCreatedColumns = 0;
+	private int $countInsertedRows = 0;
+	private int $countErrors = 0;
+
+	public function __construct(PermissionsService $permissionsService, LoggerInterface $logger, ?string $userId,
+		IRootFolder $rootFolder, ColumnService $columnService, RowService $rowService) {
+		parent::__construct($logger, $userId, $permissionsService);
+		$this->rootFolder = $rootFolder;
+		$this->columnService = $columnService;
+		$this->rowService = $rowService;
+	}
+
+	/**
+	 * @throws ContainerExceptionInterface
+	 * @throws NotFoundExceptionInterface
+	 * @throws InternalError
+	 * @throws NotFoundError
+	 */
+	public function import(int $tableId, string $path, bool $createMissingColumns = true): array {
+		if (!$this->userId) {
+			$error = 'No user in context, can not import data. Cancel.';
+			$this->logger->debug($error);
+			throw new InternalError($error);
+		}
+
+		$this->tableId = $tableId;
+		$this->createUnknownColumns = $createMissingColumns;
+
+		/** @var IConfig $config */
+		$config = Server::get(IConfig::class);
+
+		try {
+			$storageHome = $this->rootFolder->getUserFolder($this->userId);
+			if ($storageHome->nodeExists($path)) {
+				// TODO is this working if a object storage is used?
+				$completePath = $storageHome->get($path)->getPath();
+				$basePath = $config->getSystemValue('datadirectory');
+				$spreadsheet = IOFactory::load($basePath.$completePath);
+				$this->loop($spreadsheet->getActiveSheet());
+			}
```

Comment:
Ideally we could just call `$storageHome->get($path)->fopen('r')` but from a quick look at the phpspreadsheet code base it seems that it does not support reading files from a stream. Also https://github.com/PHPOffice/PhpSpreadsheet/issues/1931 is still open.

To make this work with object storage we will need a local copy of the file that is obtained through our storage abstractions. 

Something like this should work:

```suggestion
			$userFolder = $this->rootFolder->getUserFolder($this->userId);
			if ($userFolder->nodeExists($path)) {
				$file = $storageHome->get($path);
				$tmpFileName = $file->getStorage()->getLocalFile($file->getInternalPath());
				$spreadsheet = IOFactory::load($tmpFileName);
				$this->loop($spreadsheet->getActiveSheet());
			}
```